src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotParticipant.java

```diff
@@ -0,0 +1,116 @@
+ * When the transaction is aborted and changes need to be rolled back, {@link #readSnapshot} will be called
+ * to signal that the current state should revert to that of the snapshot.
+ * The snapshot object is then {@linkplain #releaseSnapshot released}, and can be cached for subsequent use, or discarded.
+ *
+ * <p>When an outer transaction is committed, {@link #readSnapshot} will not be called so that the current state of this participant
+ * is retained. {@link #releaseSnapshot} will be called because the snapshot is not necessary anymore,
+ * and {@link #onFinalCommit} will be called after the transaction is closed.
+ *
+ * @param <T> The objects that this participant uses to save its state snapshots.
+ */
+public abstract class SnapshotParticipant<T> implements Transaction.CloseCallback, Transaction.OuterCloseCallback {
+    private final List<T> snapshots = new ArrayList<>();
+
+    /**
+     * Return a new <b>nonnull</b> object containing the current state of this participant.
+     * <b>{@code null} may not be returned, or an exception will be thrown!</b>
+     */
+    protected abstract T createSnapshot();
+
+    /**
+     * Roll back to a state previously created by {@link #createSnapshot}.
+     */
+    protected abstract void readSnapshot(T snapshot);
+
+    /**
+     * Signals that the snapshot will not be used anymore, and is safe to cache for next calls to {@link #createSnapshot},
+     * or discard entirely.
+     */
+    protected void releaseSnapshot(T snapshot) {
+    }
+
+    /**
+     * Called after an outer transaction succeeded,
+     * to perform irreversible actions such as {@code markDirty()} or neighbor updates.
+     */
+    protected void onFinalCommit() {
+    }
+
+    /**
+     * Update the stored snapshots so that the changes happening as part of the passed transaction can be correctly
+     * committed or rolled back.
+     * This function should be called every time the participant is about to change its internal state as part of a transaction.
+     */
+    public void updateSnapshots(TransactionContext transaction) {
+        // Make sure we have enough storage for snapshots
+        while (snapshots.size() <= transaction.nestingDepth()) {
+            snapshots.add(null);
+        }
+
+        // If the snapshot is null, we need to create it, and we need to register a callback.
+        if (snapshots.get(transaction.nestingDepth()) == null) {
+            T snapshot = createSnapshot();
+            Objects.requireNonNull(snapshot, "Snapshot may not be null!");
+
+            snapshots.set(transaction.nestingDepth(), snapshot);
+            transaction.addCloseCallback(this);
+        }
+    }
+
+    @Override
+    public void onClose(TransactionContext transaction, Transaction.Result result) {
+        // Get and remove the relevant snapshot.
+        T snapshot = snapshots.set(transaction.nestingDepth(), null);
```

Comment:
Should we be null checking this?

Reply:
If the close callback was registered, there is necessarily a snapshot.